Bases futures on Try instead of Either

heathermiller opened the pull request:
This pull request is a result of meetings between the Scala team, the Akka team and Twitter Engineering to move toward a consolidated future implementation.

Its main contribution is the replacement of the use of `Either` with `scala.util.Try` in methods `Future.onComplete`, `Future.value`, `Future.andThen`, `Promise.complete`, and `Promise.tryComplete`.

`Try` is an abstraction adopted from the `com.twitter.util` package [[1](https://github.com/twitter/util/)]. Unlike `Either`, `Try` is projection-less, which makes it more composable and more semantically similar to futures. Furthermore, as compared to `Either`, `Try` is better suited to handling exceptions, including execption-specific combinators such as `recover` and `recoverWith` (called `rescue` in `com.twitter.util.Try`). Moreover, both fatal and non-fatal exceptions are handled consistently across combinators such as `map` and `flatMap`. 

Changes compared to Twitter's original implementation of `Try`:
- Use `scala.util.control.NonFatal` (adopted from Akka [[2](http://akka.io/)] for SIP-14) to avoid catching fatal exceptions.
- Rename `Return/Throw` (in package `com.twitter.util`) to `Success`/`Failure` (in package `scala.util`); this makes the use of `Try` in combination with `Future` more regular.
- Rename `handle` (in package `com.twitter.util`) to `recover` (in package `scala.util`).
- Rename `rescue` (in package `com.twitter.util`) to `recoverWith` (in package `scala.util`), after `scala.concurrent.Future`'s semantically-similar `recoverWith` method.
- Add `foreach` method.
- Add `orElse` method.
- Add `failed` projection.

This pull request additionally contains some small fixes (`flatMap` in `Try` now goes through the `NonFatal` extractor), clean ups, and added tests on `Try`.

@phaller has also included a `prepare` method on `ExecutionContext`s which enables abstractions to be built on top of execution contexts, such as Twitter's `Local` for handling data local to a chain of future callbacks, as previously discussed.

Review by @viktorklang, and if available @mariusaeriksen

heathermiller committed:
Basing Futures on Try instead of Either

phaller committed:
SI-6185 - add "prepare" hook to ExecutionContext

Enables important abstractions to be built on top of futures,
such as Twitter's "Local" for handling data local to a callback chain.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/70/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/773/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/70/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/773/

adriaanm commented:
one test didn't pass: jvm/future-spec

heathermiller commented:
This is weird-- it works fine, but only fails when you run partest with -optimize.

heathermiller committed:
Temporarily skips failing test due to optimizer bug SI-6188

Also swaps the arguments to method transform on Try, so as to mirror transform on scala.concurrent.Future.

heathermiller committed:
Merge branch 'try-based-futures' of https://github.com/heathermiller/scala into try-based-futures

heathermiller commented:
I think I might've discovered a bug in the optimizer. This one snuck through because instead of testing through ant (which runs the optimizer) I tend to manually run partest without -optimize.

It seems that the optimizer is incorrectly optimizing away method invocations which pass a closure containing a throw expression. I created a ticket here: https://issues.scala-lang.org/browse/SI-6188

So, seeing as this seems like an issue with the optimizer (a seemingly critical one), and not Try, I think the best course of action is to temporarily skip the failing test until the issue can be resolved with the optimizer. (Noted in the bug report).

heathermiller commented:
PLS REBUILD ALL kitteh...

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/74/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/777/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/74/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/777/

viktorklang commented on src/library/scala/concurrent/ExecutionContext.scala:
```diff
@@ -27,6 +28,13 @@ trait ExecutionContext {
+   *  execution context which should be used to schedule the execution
+   *  of the task associated with `f`.
+   */
+  def prepare[T, U](f: Try[T] => U): ExecutionContext =
```
I think this is overly specific, ExecutionContext shouldn't be tied into Try.

```
def bind[T, R](f: T => R): ExecutionContext = this
```

viktorklang replied:
Any comment here?

heathermiller replied:
This is for @phaller -- hoping he'll chime in. (Will ping him).

phaller replied:
I agree that it's not necessary to be so specific. OK, so then let's just make it generic.

```
def prepare[T, U](f: T => U): ExecutionContext = this
```

I'm not sure about renaming it. `bind` makes sense, but it's one of those very, very general names. In the meeting we always used `prepare` and intuitively it makes sense to me.

Anyway, Heather can add this change to the commit with the fixes.

viktorklang replied:
Deal

mariusae replied:
We might even drop the argument altogether. I don’t think access to the actual function is actually needed:

``` scala
def prepare(): ExecutionContext = this
```

viktorklang replied:
@mariusaeriksen That would indeed be even better!

heathermiller replied:
Sure thing, sounds good.

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -308,8 +308,8 @@ trait Future[+T] extends Awaitable[T] {
+      case Success(v) =>
         try {
           if (pred(v)) p success v
           else p failure new NoSuchElementException("Future.filter predicate is not satisfied by: " + v)
```
We should probably not do a toString on "v" here as it may be a very large object, right?

heathermiller replied:
Let's keep the review of the pull request focused on this pull request (i.e. on what has been changed), not on general details/fixes/cleanups for scala.concurrent. Instead, why not make a list of last minute clean up items like these, and make a separate pull. Else this pull request will no longer be described by its title.

viktorklang replied:
Did I say that it should be addressed in this PR? Do you disagree with my comment?

heathermiller replied:
I don't disagree with your comment. I'm just trying to keep records and commentary in pull requests clean for future reference. Thanks for understanding!

viktorklang commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -141,24 +146,25 @@ private[concurrent] object Promise {
-      case Right(r) => r
+      case Failure(e)  => throw e
+      case Success(r) => r
     }
```
Doesn't Try.get throw the exception? if so, you can avoid this patmat

heathermiller replied:
Sure, thanks for catching it. Fixed in forthcoming commit.

viktorklang commented on test/files/jvm/future-spec/TryTests.scala:
```diff
@@ -0,0 +1,119 @@
+        Failure[Int](e) flatMap(x => Success(1 + x)) mustEqual Failure(e)
+      }
+
+      // "when there is an exception" in {
```
Should be a comment explaining why this is commented

heathermiller replied:
There is. Read the discussion.

heathermiller replied:
It uncovered a bug in the optimizer, I opened a ticket this morning. https://issues.scala-lang.org/browse/SI-6188

viktorklang replied:
Is it possible to not comment the tests out, but making so that we test that it fails with -optimize, so the guy who fixes SI-6188 also can fix the test?

heathermiller replied:
The ticket already has a minimized test case, unrelated to Try. The minimized test case will definitely go into the test suite once the bug is fixed, so we shouldn't have to reorganize the Try tests (which means splitting into separate files, etc). However, this specific test file is also documented in the bug report, so it'd also be uncommented when the bug is fixed. But as pointed to above, as this bug is completely unrelated to Try and Futures, and is purely a bug related to the optimizer, so this pull request shouldn't be held up due to it. I should also point out that the bug is labeled as critical for the time being, meaning it should hopefully come with a fix soon.

viktorklang replied:
I agree with your assessment, the ticket is very solid and the link to what test need to be reenabled is very clear.

viktorklang commented on test/files/jvm/future-spec/TryTests.scala:
```diff
@@ -0,0 +1,119 @@
+      }
+    }
+  }
+}
```
Tests are missing for atleast: andThen, transform (potentially more)

heathermiller replied:
`andThen` is an alias for `flatMap`, so I think we're covered on tests for that. I'll add a `transform` test to try-type-tests.scala.

viktorklang replied:
On second thought I think andThen should not be there. We don't have any precedent for this kind of duplication, right?

heathermiller replied:
https://github.com/twitter/util/blob/master/util-core/src/main/scala/com/twitter/util/Try.scala#L133
I'm unsure of removing it, mostly because I'm not completely certain why it was added in the first place. I imagine that it improves readability in some important use case at Twitter (as, in general, uses of `flatMap` can get obscure at times). Maybe @mariusaeriksen could chime in? In any case, it's quite harmless, and perhaps it's nicer and clearer to read in application code-- that merit might outweigh the mere actuality that it's an alias.

mariusae replied:
I would actually have not created this alias if I were to do it again—I'd encourage removing it. Initially we thought of it as a more natural way to express sequencing (this and then that) but everybody got quickly used to `flatMap` and what it actually meant — to the point that seeing `andThen` in any code is quite jarring.

heathermiller replied:
Alrighty then-- game, set, match for the eradication of `andThen`.

viktorklang commented:
I think toOption and toSeq don't belong on Try, it's mixing concerns and one could argue why there is no toSet etc.
I suggest having the "to"-method that was recently added to the collections so you could transform a Try to essentially any container.

Also, line 209 in Try.scala (https://github.com/scala/scala/blob/1241e40ed4442abd635f44ce997013268b956b4b/src/library/scala/util/Try.scala#L209)
catches Throwable, should be NonFatal(t)

heathermiller commented:
In response to @viktorklang's comment about on line 209 on Try.scala being Throwable and not NonFatal, that's not true. You linked to an individual commit made days ago that has been corrected in this pull request. Look at the diff in the pull request, line 193 on Try:
https://github.com/scala/scala/pull/1056/files#L5R193

viktorklang commented:
Doh! I was comparing against the latest nightly. Great work Heather!

heathermiller commented:
No worries-- getting lost in random commits on github is all too common.

In response to the other comment about `toSeq` and `toOption`, as mentioned earlier, for the sake of clean history, we should address this concern and any other clean ups or tweaks, documentation fixes, etc, in a separate pull request (rather than adding more commits to this one). 

However, I do agree, it'd be cleaner to remove `toSeq` in favor of something more generic like the `to` method recently introduced to collections, sure. Though I'm not sure that `to` works with `Option` so maybe we should just remove `toSeq` all together and leave `toOption`? Twitter also has a `toOption` on their Try, maybe we should try to keep parity with their Try? Worst case, later, if it's even possible, we can reimplement `toOption` in terms of `to` leaving the interface.

viktorklang commented:
Yeah, I'm against toX methods in general as they are hard-coded and sort of gives special treatment to other types.
Might be more powerful to add "fold"?

```
def fold[U](success: T => U, failure: Throwable => U): U

toOption: try.fold(Some(_), _ => None)
toSeq: try.fold(List(_), _ => Nil)
```

mariusae commented on src/library/scala/util/Try.scala:
```diff
@@ -54,6 +54,7 @@ import language.implicitConversions
  *
  * `Try` comes to the Scala standard library after years of use as an integral part of Twitter's stack.
  *
+ * @author based on Marius Eriksen's original implementation in com.twitter.util.
```
I would just say “Twitter‘s” - there were many authors of Try.

heathermiller replied:
Ok, sure thing.

heathermiller committed:
Added tests, removal of unnecessary methods, fixes prepare

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/115/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/819/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/115/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/819/

jsuereth commented:
failure looks valid.

heathermiller committed:
Doc fix on exec ctx prepare method, fix to tests

heathermiller commented:
Yep, that earlier one was a premature commit. Everything brought up by the reviewers should now be addressed.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/120/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/824/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/120/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/824/

phaller commented:
Aborted because I realized that a test I added to scala-concurrent-tck.scala which tests the `prepare` method would fail. Will correct this momentarily, otherwise it LGTM.

phaller committed:
Fix test for ExecutionContext.prepare

heathermiller commented:
PLS REBUILD ALL kitteh... kthx

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/127/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/831/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/127/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/831/

heathermiller commented:
So all should now be addressed. Señor @viktorklang, can I a LGTM?

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -308,11 +308,11 @@ trait Future[+T] extends Awaitable[T] {
         try {
           if (pred(v)) p success v
-          else p failure new NoSuchElementException("Future.filter predicate is not satisfied by: " + v)
+          else p failure new NoSuchElementException("Future.filter predicate is not satisfied")
```
excellent

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -377,22 +377,15 @@ trait Future[+T] extends Awaitable[T] {
-        }
-      case otherwise => p complete otherwise
-    }(executor)
+    onComplete { case tr => p.complete(tr recover pf) }(executor)
```
Excellent

viktorklang commented on src/library/scala/concurrent/impl/Future.scala:
```diff
@@ -21,7 +21,7 @@ private[concurrent] object Future {
     override def run() = {
       promise complete {
-        try Right(body) catch { case NonFatal(e) => Left(e) }
+        try Success(body) catch { case NonFatal(e) => Failure(e) }
```
```
promise complete Try(body) ?
```

viktorklang commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -15,42 +15,46 @@ import scala.concurrent.{ ExecutionContext, CanAwait, OnCompleteRunnable, Timeou
     require(value eq null) // can't complete it twice
     value = v
+    // Note that we cannot prepare the ExecutionContext at this point, since we might
+    // already be running on a different thread!
```
Great comment

viktorklang commented:
Aside from my comments, LGTM

adriaanm commented:
@viktorklang could you summarize the blockers for your giving an outright LGTM on this pull request?

I would really like to move forward on this, thus I propose anything that's not truly critical be moved to a separate pull request that builds on this one.

viktorklang commented:
@adriaanm No blockers.

adriaanm commented:
cool, thanks everyone!

heathermiller commented:
Great, thanks a lot @mariusaeriksen and @viktorklang!!

Blaisorblade commented:
@heathermiller, even as an outsider, I have a problem with putting infos on commits in the pull request instead of the commit - in short, this info is not there when somebody looks for it.
What you write in the pull request text is unavailable and not reachable when you `git blame` a change and I'm looking at https://github.com/heathermiller/scala/commit/3cb0e784a05db7d0b542cec9bf4c5fbf3772a6cf.
This bit me a couple days ago while investigating https://issues.scala-lang.org/browse/SI-6190. I found this pull request by accident and only today.

Most of the pull request text should IMHO just be part of the commits themselves (`git commit --amend` and `git rebase` allow doing that), unless Scala's workflow is intentionally different.
